tinygrad/realize.py

```diff
@@ -172,6 +172,40 @@ def _is_padding_okay(buf:LazyBuffer, realizes:Set[LazyBuffer]) -> bool:
   if buf.op in UNSAFE_PAD_OPS: return False
   return all(_is_padding_okay(x.base, realizes) for x in buf.srcs)
 
+def optimize(schedule: List[ScheduleItem]) -> List[ScheduleItem]:
+  # find assigns, map positions of output LazyBuffers, init graph
+  assigns: Dict[Buffer, int] = {}
+  positions: Dict[LazyBuffer, int] = {}
+  graph: Dict[int, Set[int]] = {}
+  for i,si in enumerate(schedule):
+    graph[i] = set()
+    positions[si.out] = i
+    if si.out.output_buffer: assigns[si.out.output_buffer] = i
+  # Figure out dependencies
+  for i,si in enumerate(schedule):
+    for inp in si.inputs:
+      if (pos := positions.get(inp)) is not None:
+        graph[pos].add(i)
+      if inp.realized and (pos := assigns.get(inp.realized)) is not None and pos != i:
+        graph[i].add(pos)
+  del assigns, positions
+
+  top_order: List[int] = []
+  in_degrees: Dict[int, int] = {k:0 for k in graph.keys()}
+  for n in flatten(graph.values()): in_degrees[n] += 1
+  free: Deque[int] = deque([idx for idx,deg in in_degrees.items() if deg == 0])
+
+  while free:
+    node = free.popleft()
```

Comment:
I am thinking about this logic, while it's looks simple and good on some schedules, but is it generally a good heuristic for the thing we trying to solve? In my understanding we want to hide transfer times. This does really primitive reordering, how efficient is it with other schedules?

Reply:
14.88 tok/s on llama 7b on 6 gpus (w/o ring, just this optimization) vs 15.10 on #3626

Reply:
```
(.venv) uuuvn@tiny10:~/src/tinygrad$ STEPS=10 GPUS=6 BS=720 HSA=1 python3 examples/hlb_cifar10.py
shuffling training dataset in 1287.12 ms (epoch=0)
  0 14875.71 ms run, 14871.37 ms python,    4.34 ms HSA * 6, 1198.74 loss, 0.001495 LR, 1.10 GB used,     63.62 GFLOPS,    946.39 GOPS
  1 1131.90 ms run, 1129.36 ms python,    2.54 ms HSA * 6, 1198.97 loss, 0.002989 LR, 6.68 GB used,    834.47 GFLOPS,    944.53 GOPS
  2   46.13 ms run,   13.32 ms python,   32.80 ms HSA * 6, 2044.32 loss, 0.003133 LR, 6.69 GB used,  20476.73 GFLOPS,    944.53 GOPS
  3   38.37 ms run,   11.75 ms python,   26.61 ms HSA * 6, 2233.89 loss, 0.002698 LR, 6.69 GB used,  24617.35 GFLOPS,    944.53 GOPS
  4   38.32 ms run,   11.66 ms python,   26.66 ms HSA * 6, 1795.34 loss, 0.002262 LR, 6.69 GB used,  24645.85 GFLOPS,    944.53 GOPS
  5   38.27 ms run,   11.69 ms python,   26.59 ms HSA * 6, 1633.46 loss, 0.001827 LR, 6.69 GB used,  24679.91 GFLOPS,    944.53 GOPS
  6   38.61 ms run,   11.63 ms python,   26.97 ms HSA * 6, 1630.74 loss, 0.001392 LR, 6.69 GB used,  24466.32 GFLOPS,    944.53 GOPS
  7   38.44 ms run,   11.67 ms python,   26.77 ms HSA * 6, 1364.76 loss, 0.000956 LR, 6.69 GB used,  24572.66 GFLOPS,    944.53 GOPS
  8   38.51 ms run,   11.59 ms python,   26.92 ms HSA * 6, 1406.93 loss, 0.000521 LR, 6.69 GB used,  24525.51 GFLOPS,    944.53 GOPS
  9   38.40 ms run,   11.70 ms python,   26.70 ms HSA * 6, 1381.90 loss, 0.000086 LR, 6.69 GB used,  24594.88 GFLOPS,    944.53 GOPS
shuffling test dataset in 123.60 ms (epoch=0)
eval     1376/10080 13.65%,    2.49 val_loss STEP=10 (in 1206.81 ms)
```
hlb_cifar

Reply:
I mean results look good. But if we get a different initial schedule, is the heuristic still as efficient? If we have 4 available kernels to execute and last of them has a long copy child, this won't hide this long copy time, since it will execute other kernels, before execution 4th to start copy in parallel.